feat(merge): Added 'mergeDriver' parameter

araknast opened the pull request:
Adds an optional `onMergeConflict` parameter to `merge()`. 
The parameter is a callback which takes the string contents of two conflicting files as input and returns the merged text. If no such parameter is provided, and `dir` is set when calling merge, markers are added to the conflicting files in the working directory and a `MergeNotSupportedError()` is thrown, aborting the merge.

Removes configuration for marker size and output type with their corresponding tests, as these are now redundant. 

Should fix #841 and #325 

## I'm adding a parameter to an existing command X:

- [x] add parameter to the function in `src/api/X.js` (and `src/commands/X.js` if necessary)
- [x] document the parameter in the JSDoc comment above the function
- [x] add a test case in `__tests__/test-X.js` if possible
- [x] if this is your first time contributing, run `npm run add-contributor` and follow the prompts to add yourself to the README
- [ ] squash merge the PR with commit message "feat(X): Added 'bar' parameter"

jcubic commented:
Thanks for the PR with a fix, will check it on Monday.

jcubic commented on __tests__/test-merge.js:
```diff
@@ -564,7 +569,135 @@ describe('merge', () => {
+        theirName = 'theirs',
+      }) => {
+        const LINEBREAKS = /^.*(\r?\n|$)/gm
+        const ours = ourContent.match(LINEBREAKS)
```
What this line is doing? It's hard to understand. Is it the same as `ourContent.split(/\r?\n/)`?

araknast replied:
This regex is copied from the original `mergeFile` implementation, but yes it's essentially equivalent to `ourContent.split(/\r?\n/)`

jcubic commented:
How about if there is more than one conflicted file? This should be handled as well. I think that the callback function (event) should be called on each file also it should return a promise. Or there should be an exception thrown and the user should call the function resolve error.

This should work so I can make this flow work like in canonical git:

merge -> conflict -> manual edit the file -> continue merge.

I would prefer the API that spits conflict and resolver into two functions for example:

```javascript
git.merge({...args, conflictMarkers: true});
// manual edit
git.resolve({...args});
```

There can be two APIs one with a callback function (event) and one when the user can code and do manual resolving of the conflict.

jcubic commented on __tests__/test-mergeFile.js:
```diff
@@ -29,33 +29,4 @@ describe('mergeFile', () => {
-    })
-    expect(cleanMerge).toBe(false)
-    expect(mergedText).toEqual(await fs.read(`${dir}/aoc3-m10.txt`, 'utf8'))
-  })
```
If you removed those two tests does it mean that this API no longer works? This would be a breaking change.

araknast replied:
Options for diff3 and custom marker size were only exposed internally, and never used by anything except these tests, so this wouldn't be a breaking change.

araknast commented:
> How about if there is more than one conflicted file? And you need to do different action than just prefer ours or theirs. This should be handled as well.
The `prefer ours` and `prefer theirs` tests are just simple examples meant to check that `onMergeConflict` is implemented properly. The user could implement any number of strategies provided they are the same regardless of file. If the user wants to resolve conflicts differently for different files they can forgo specifying `onMergeConflict`, allow the files in the working directory to be marked, resolve the conflicts manually and then add and commit. Alternatively, we could add a `filepath` parameter to `onMergeConflict` so the conflict resolver could take different actions for different files.

> This should work so I can make this flow work like in canonical git:
> 
> merge -> conflict -> manual edit the file -> continue merge.
This is essentially the behavior when `onMergeConflict` is not specified. After the error is thrown, the user can edit the marked files in the working directory, add and commit just like in canonical git. I was incorrect in stating a `MergeNotSupportedError()` aborts the merge, it simply interrupts the merge, and writes all merged and conflicting files to the working directory similarly to canonical git. 

> There can be two APIs one with a callback function (event) and one when the user can code and do manual resolving of the conflict.
I'm not sure how this is different from what's implemented here. `merge({...args, onMergeConflict})` essentially functions as the callback api while `merge({...args})` functions at the manual resolution api. Instead of 
```
git.merge({...args, conflictMarkers: true});
// manual edit
git.resolve({...args});
```
We do
git.merge({...args})
.catch(e => {
  if (!e instanceof MergeNotSupportedError) throw e;
})
// manual edit
git.add({...args})
git.commit({...args})
```
Although I do think we need a new type of error to throw, e.g. `MergeConflictError()` to differentiate merge conflicts that can be resolved manually from those that can't. 

In the former example, assuming `git.merge()` doesn't throw an error when a conflict occurs and `conflictMarkers` is `true`, I'm not sure how we determine when a merge is interrupted and therefore needs manual resolution. Assuming `git.merge()` does throw an error, the flow is the same except we just use the existing `git.add()` and `git.commit()` instead of adding a new `git.resolve()`.

jcubic commented:
Ok, thanks for your work and explanation. Will this work?

```javascript
git.merge({...args, conflictMarkers: true})
.catch(e => {
  if (e instanceof MergeConflictError) {
     // conflict marker exists I can edit and resolve conflicts by hand
     
  }
});

git.merge({...args, conflictMarkers: false})
.catch(e => {
  if (e instanceof MergeConflictError) {
     // conflict happen merge aborted
     
  }
});
```

It would be nice to see unit tests for this scenario.

Another missing feature is the abort of the merge. If we have conflicts and the merge was stopped in the middle we need to be able to abort the merge.

araknast commented:
This flow should now be possible with the `clean` parameter, e.g:

```
git.merge({...args})
.catch(e => {
  if (e instanceof MergeConflictError) {
     // conflict marker exists I can edit and resolve conflicts by hand
  }
});

git.merge({...args, clean: true})
.catch(e => {
  if (e instanceof MergeConflictError) {
     // conflict happen merge aborted - working and stage area clean
  }
});
```
Implementing merge abortion is a bit trickier, as it seems the `GitIndex` structure isomorphic-git currently uses is implemented using `Map()`, so since each file path is permitted only one entry in the index, we can't have the index stages required for proper `merge --abort` behavior. I'll see what I can do, hopefully it's fairly trivial to fix.

araknast commented:
I agree, I think `abortOnConflict` works better, and tests should be working now that I've brought my repo up to date with upstream. 

I've gone ahead and added some documentation for the new parameter, and also changed the default behavior to throw an error by default to match with what we do currently.

araknast commented:
@jcubic since we are leaving merge --abort as a separate issue, is there anything else needed to get this merged?

jcubic commented:
No, sorry I need to look through your PR again since you've added a lot of commits. I simply didn't have time to check the code again.

jcubic commented on src/utils/mergeFile.js:
```diff
@@ -2,26 +2,27 @@ import diff3Merge from 'diff3'
   // Here we let the diff3 library do the heavy lifting.
   const result = diff3Merge(ours, base, theirs)
 
+  const markerSize = 7
```
I would not get rid of this option and use:

```javascript
 mergeFile({ branches, contents, markerSize = 7 }) {

})
```

Just in case, so we don't remove something that may be needed.

jcubic commented:
I've added one comment, why did you remove markerSize? It seems that it may be useful to not remove functionality unless there is a good reason to remove that option.

The second thing is can you add both mergeDrives codes from unit test to documentation? Prefer ours and prefer theirs.

jcubic commented:
Ok, so maybe there should be just one unit test and one documentation piece that show how to produce conflict like in canonical git where you create the markers yourself in the driver. So you can edit the files and add a merge commit. I'm not sure about merge commit but there should be a state where there is conflict in files with markers.

jcubic commented:
And if you think that `markerSize` is not needed you can remove it, unless it will work when you don't provide the MergeDriver and you will end up with conflict in files.

araknast commented:
Ok, I've added tests and documentation for manual resolution of merge conflicts, and re-removed the markerSize parameter. The documentation also explains the current limitations of the merge command, and how to emulate the behavior of `git merge --abort` and `git merge --continue` since we don't have a proper index model yet.

jcubic commented:
But what about markers in the output files? You just added all files and commit after conflict.
You've said that there will be markers that users can add in mergeDriver. How he can edit the files and resolve the conflicts if there are no markers?

araknast commented:
If no merge driver is specified, the default merge driver (mergeFile) is used, which automatically adds markers to the files with conflicts.

jcubic commented:
You've said that makerSize is not needed because you write them by hand. Now you say that to have a conflict with markers you don't use the driver. If that is the case then you need to add back markerSize so you can have the same output as before.

araknast commented:
If the user wanted to change the marker size they can simply copy the code from `mergeFile` and change the `markerSize` variable there, e.g.

```
const mergeDriver = ({ contents, branches }) => {
  const ourName = branches[1]
  const theirName = branches[2]

  const baseContent = contents[0]
  const ourContent = contents[1]
  const theirContent = contents[2]

  const ours = ourContent.match(LINEBREAKS)
  const base = baseContent.match(LINEBREAKS)
  const theirs = theirContent.match(LINEBREAKS)

  const result = diff3Merge(ours, base, theirs)

  // const markerSize = 7
  const markerSize = 28 // change marker size

  let mergedText = ''
  let cleanMerge = true

  for (const item of result) {
    if (item.ok) {
      mergedText += item.ok.join('')
    }
    if (item.conflict) {
      cleanMerge = false
      mergedText += `${'<'.repeat(markerSize)} ${ourName}\n`
      mergedText += item.conflict.a.join('')

      mergedText += `${'='.repeat(markerSize)}\n`
      mergedText += item.conflict.b.join('')
      mergedText += `${'>'.repeat(markerSize)} ${theirName}\n`
    }
  }
  return { cleanMerge, mergedText }
}

git.merge({...args, mergeDriver})
.catch(e => {
  if (!(e instanceof MergeConflictError)) throw e
})
// manual resolution
git.add({...args})
git.commit({...args})
```

Should I add this to the documentation?

jcubic commented:
Yes, it would be nice to have this in the documentation.

araknast commented:
Ok, I've added an example to the mergeDriver documentation.

jcubic commented:
Thank you
